### DISCUSSION TITLE
Windows status

### DISCUSSION THREAD

**Discussion by servantftransperfect** [AcademySoftwareFoundation/rez#1288] [Q&A]
Hello,

Is there some place to find information on rez for windows ?

I see various very old discussions and PRs.

I tried on my computer and it doesn't work as is (not finding nmake even using the vs command line).

Is there something working somewhere ? Are there some planned release ?

Thanks

Fabien

**Comment by instinct-vfx** [comment]
Hey Fabien,

rez does work and support Windows. There are some gotchas on Windows and Documentation is not necessarily reflecting these properly as of yet.

Particularly handling Paths and Environment Variables can be tricky with rez. That can be very unintuitive but has good reasons. Could you explain in a bit more detail what you want to achieve, what you tried and how it failed?

Cheers,
Thorsten

**Reply by servantftransperfect** [reply]
please see below

**Comment by servantftransperfect** [comment]
I tried to make something as simple as possible : 

``` 
CMAKE_MINIMUM_REQUIRED(VERSION 2.8)

include(RezBuild)
include(ExternalProject)

set(url_eigen https://gitlab.com/libeigen/eigen.git)
set(tag_eigen 3.4)


ExternalProject_add(
    eigen
    GIT_REPOSITORY url_eigen
    GIT_TAG tag_eigen
    PREFIX eigen
    CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX}
)
```

```
`name = "eigen"

version = "3.4"

authors = [
]

description = \
    """
    """

variants = [
]

tools = [
]

uuid = "repository.eigen"


def commands():
    pass

    if building:
        pass
```

**Comment by Jawabiscuit** [comment]
Hi @servantftechnicolor, I made [msvc](https://github.com/Jawabiscuit/rez-packages/tree/main/msvc) and [vs](https://github.com/Jawabiscuit/rez-packages/tree/main/vs/15) when I encountered the same roadblocks on Windows. Those issues manifest from the fact that rez doesn't come with batteries included to build cmake projects. You need visual studio and cmake to be available in the rez environment and it seems on Windows, at least for the time being, the way to get those in the environment is for them to be provided by packages.

The msvc (14.16) package is essentially the same as vs (15) with the only major differences being the former provides just the basic build tools and the latter additionally provides `devenv` UI. This is so that you can have a smaller package if you only intend to use the package for it's building capability (msvc) and not it's interactive debugging capability (vs), among other less significant things.

These packages only provide the bare minimum of build tools to compile C++ using nmake, cl, etc..., and provide the built-in cmake. The installer, for whichever package you might choose, you must download yourself from the Microsoft website. Shortcut link is included in the README. They use `pwsh` to build so that should be the only requirement.

Maybe if you're up to the task, you could give any one of those packages a try. I've had much success using them to build practically all my projects, Python or C++. I love how simple it is to build Python projects especially. When you have downloaded and built a package, just include it as one of your `private_build_requires` in the package you wish to build with cmake and you should be off to the races!

**Reply by instinct-vfx** [reply]
As a small correction, rez comes with all batteries included for CMake, but not for Visual Studio. You can run Cmake fine from a build command for example. The environment of a visual studio prompt is set for that very session and some of these (mostly PATH) are taken from the registry in rez (and there are good reasons to do so).

**Reply by servantftransperfect** [reply]
Thanks a lot for those answers. I'll give it a try and i'll come back to write my observations.

**Reply by Jawabiscuit** [reply]
Awesome!

A couple things I'd like to disclose which might not be immediately obvious if you're going to try out msvc or vs packges is the packages payload the build tools, so it doesn't rely on any existing visual studio install and it gets installed into the package itself. This has some side-effects and I'd like to know more about to better guage if payloading is a good or bad idea. These side-effects are probably only going to be noticable in multi-user studio environments.

1. I'm thinking package caching, which I don't use on my home machine but would like to know more about, may need investigation. 
2. A known issue with using these packages is that windows sdk installation isn't marshalled and payloaded and cannot be at the moment but it is on the roadmap for Microsoft IIRC. In effect, the sdk dir is created by the build tools installer in its default location. 
3. I haven't run a full "rez release" of them so I don't know if there are any side effects to that either.

I would love to know answers to these questions as well as other's experience! They have been working great so far with my single-user development workflows.

**Reply by bpabel** [reply]
For whatever reason, the default `cmake` build system for rez on windows is set to `nmake` (which requires visual studio), instead of `make`, like it is on mac/linux. You can set it back to `make` in your `rezconfig.py`, which is much easier to get working:

```python
plugins = {
    "build_system": {
        "cmake": {
            "build_system": "make",
        },
    },
}
```

**Reply by Jawabiscuit** [reply]
Great, I've been wondering how to do this or if it was possible! Been trying command line args which don't seem to have an effect.

**Reply by ColinKennedy** [reply]
FWIW I ran into this a while back and found there's a CLI option to force the build to use make instead of nmake. (`--cmake-build-system`, or `--bs`, is [located here](https://github.com/nerdvegas/rez/blob/6bd2f451e2ca464abcdcca3c7060a81fca1c2765/src/rezplugins/build_system/cmake.py#L89))) But it's better to put it in your rezconfig.py, so it you don't need to remember to do it each time

**Reply by Jawabiscuit** [reply]
That information would be really useful if outputted if executing `rez build --help`. Also, the `build_systems` dict keys. I guess I will be diving into the code more often, but this really is an example of something that can be improved to help new users. I still consider myself one after having used it for a year, there's that much to learn.

I believe I tried adding args after `--` like `-DCMAKE_GENERATOR` which didn't have the effect I was looking for, but IDK maybe an additional `--` was necessary. I didn't spend too much time searching for an answer thinking I'd come back to it with a fresh perspective, but it seemed like rez was doing something to always override it to use nmake makefiles generator.

If this additional option help could be addressed it would save some headbanging for those who don't have the bandwidth to study the source code, and obviously diminish the challenges of using cmake on windows. Documenting this option as well as @bpabel 's answer as a best practice  for getting started on windows would undoubtedly be added bonuses.

Maybe the `--bs` option could be renamed to `--cb --child-build-system` or `-g --build-generator` to avoid confusion with the `-b --build-system` that's already being used?

**Reply by instinct-vfx** [reply]
The reason rez-build does not output this is because cmake is a plugin, and rez build only outputs help for itself. Would be nice to have build systems support custom help definitely though. You can use `rez -i` to output a list of all plugins, that also includes build systems.

**Reply by Jawabiscuit** [reply]
Good to know, thanks.  As you said, it seems imperative to me that they'd also output their help to the terminal. It is a plugin, but it feels like it's part of rez. If I recall, there's impetus to split plugins off into separate repos though? That usage of `rez i` is also unintuitive in my opinion. Maybe a `--list-plugins` would be better?

**Reply by ColinKennedy** [reply]
Normally for plug-in CLI things, I'd split it as a separate subparser. e.g. `rez-build -ci` would do its regular build logic but if provided a positional, `rez-build cmake --help --other-args` it'd show the cmake plugin's help. And `rez-build --help` would show all available subparser names. Seems like that'd maintain backwards compatibility

**Reply by instinct-vfx** [reply]
I am not so sure about sub-parsers vs. parameters. Need to think about that some more

**Reply by Jawabiscuit** [reply]
> Hi @servantftechnicolor, I made [msvc](https://github.com/Jawabiscuit/rez-packages/tree/main/msvc) and [vs](https://github.com/Jawabiscuit/rez-packages/tree/main/vs/15) when I encountered the same roadblocks on Windows. Those issues manifest from the fact that rez doesn't come with batteries included to build cmake projects. You need visual studio and cmake to be available in the rez environment and it seems on Windows, at least for the time being, the way to get those in the environment is for them to be provided by packages.

In efforts to make this information more accessible to new Windows users and to set the record straight, I wanted to report at long last I was able to reprise my effort in finding a simpler solution for using cmake on Windows with a fresh install of rez. I figured out one way is to make simple rez packages for cmake and make and override the child build system for cmake (thanks to hints provided in this discussion) and change it to make. That isn't the end of the story, but it was were I left off a while back. For whatever reason on Linux I've never had to do this, but then again I never set it up from scratch on Linux, but also setting `project(foo_project NONE)` in CMakeLists.txt tells cmake the project isn't compiled. I knew that since diving into cmake a while back but have been out of the practice since you don't need to use `project` for rez packages.

To recap:

- It isn't absolutely necessary to include a Visual Studio in the environment, as a package or otherwise, as I had been as a workaround unless your project is compiled.
- Create and build packages for cmake and make.
    - Alternative tools that preclude the necessity for making packages exist in rez, but in the long-run you might experience undesireable solves: [parent_variables](https://github.com/AcademySoftwareFoundation/rez/wiki/Configuring-Rez#parent_variables), [all_parent_variables](https://github.com/AcademySoftwareFoundation/rez/wiki/Configuring-Rez#all_parent_variables), [setenv](https://github.com/AcademySoftwareFoundation/rez/wiki/Package-Commands#setenv), [unsetenv](https://github.com/AcademySoftwareFoundation/rez/wiki/Package-Commands#unsetenv)
- Include `project(foo_project, NONE)` near the top of CMakeLists.txt
- Execute `rez-build --bs make` and optionally include `-ic` at the end

That's it! 

P.S. I'm in the process of updating packages I've shared for [cmake](https://github.com/Jawabiscuit/rez-cmake) and [make](https://github.com/Jawabiscuit/rez-make) for Windows. The make package I'm sharing requires better support for gitbash on Windows. Hopefully this [PR](https://github.com/AcademySoftwareFoundation/rez/pull/1364) makes that a reality. 🙏

**Comment by nerdvegas** [comment]
The thing is though, there's no help to show for the cmake plugin (for eg)
beyond cmake's own help. Rez is just allowing you to pass args as a
passthrough.

**Reply by instinct-vfx** [reply]
That's a good point, but not necessarily true for all plugins, and i think there would be value in supporting plugin help to be presented at the respective rez commands. That said it feels like this might be a 3.x topic.
